// @fb/firestore/domains/index.js
import Domains, * as _ from './Domains.svelte'

export const registerDomain = _.registerDomain
export const unregisterDomain = _.unregisterDomain
export const duplicatePage = _.duplicatePage
export const checkIfUrlTaken = _.checkIfUrlTaken
export const setPagePassword = _.setPagePassword
export const updatePage = _.updatePage
export const setPage = _.setPage
export const changePageUrl = _.changePageUrl
export const deleteComponentFromDomain = _.deleteComponentFromDomain
export const saveSymbolToDomain = _.saveSymbolToDomain
export const updateInstancesInDomain = _.updateInstancesInDomain
export const getComponents = _.getComponents
export const addSiteCollaborator = _.addSiteCollaborator
export const acceptInvitation = _.acceptInvitation
export const registerSubdomain = _.registerSubdomain
export const deregisterSubdomain = _.deregisterSubdomain
export const setSubdomainRepo = _.setSubdomainRepo
export const getDomainData = _.getDomainData
export const getAllData = _.getAllData
export const getPageData = _.getPageData
export const createPage = _.createPage
export const deletePage = _.deletePage
export const getPagesInSubdomain = _.getPagesInSubdomain
export const checkIfSubdomainExists = _.checkIfSubdomainExists
export const saveToSite = _.saveToSite
export const getAllPages = _.getAllPages
export const saveAllPages = _.saveAllPages
export const getSymbol = _.getSymbol
export const getAllSymbols = _.getAllSymbols
